src/timely-util/src/probe.rs

```diff
@@ -150,3 +153,38 @@ impl<T: Timestamp> Clone for Handle<T> {
         }
     }
 }
+
+/// Creates a stream that flows progress updates from a probe.
+///
+/// The returned stream is guaranteed to never yield any data updates, as is reflected by its type.
+// TODO: Replace `Infallible` with `!` once the latter stabilizes.
+pub fn source<G, T>(scope: G, name: String, handle: Handle<T>) -> Stream<G, Infallible>
+where
+    G: Scope<Timestamp = T>,
+    T: Timestamp,
+{
+    let mut builder = AsyncOperatorBuilder::new(name, scope);
+    let (_output, output_stream) = builder.new_output();
+
+    builder.build(move |capabilities| async move {
+        let mut cap_set = CapabilitySet::from(capabilities);
+
+        let mut downgrade_capability = |frontier: AntichainRef<T>| {
+            // Clone the `CapabilitySet` before attempting to downgrade. The probe's frontier might
+            // jump backward, causing the downgrade to fail, which could leave the `CapabilitySet`
+            // in an inconsistent state. We want to revert to the previous version in this case.
+            let mut new_cap_set = cap_set.clone();
+            if new_cap_set.try_downgrade(&frontier).is_ok() {
+                cap_set = new_cap_set;
+            }
+
+            !frontier.is_empty()
+        };
+
+        while handle.with_frontier(&mut downgrade_capability) {
+            handle.progressed().await;
+        }
```

Comment:
That is an interesting way of expressing a `do { .. } while (..)` loop!  You can keep around an `Antichain` and avoid the `try_downgrade` footgun:

```rust
let mut cap_set = CapabilitySet::from(capabilities);
let mut frontier = Antichain::from_elem(T::minimum());
loop {
    handle.with_frontier(|f| {
        if PartialOrder::less_than(&frontier.borrow(), &f) {
            frontier = f.to_owned();
        }
    });
    cap_set.downgrade(&frontier.borrow());
    handle.progressed().await;
}
```

Reply:
Updated to use the cached-frontier approach.

I didn't move to a `loop` like you suggested though, because that makes it hard to break out of it when the frontier has advanced to the empty frontier. And I think if we don't break the loop, the operator will be stuck `await`ing and never be cleaned up.